### PR TITLE
[Regtest][Tests][RPC] Regtest mode + Test suite.

### PR DESCRIPTION
**Pull request by furszy**
#### This PR contains a fully working regtest + new test suite, including:

* On-Demand PoW blocks generation.
* On-Demand  PoS blocks generation.
* Zerocoin block height activation.
* On-Demand zPoS blocks generation.
* New test suite for the called "Fake Stake" vulnerability, functional test coverage for #803 expanded.
* Several functional tests ported and working.
* Creation of new rpc commands (waitforblock, waitfornewblock, waitforblockheight, createrawzerocoinstake).

#### "Fake Stake" Test Suite.

The suite is contained in the folder test/functional/fake_stake.
There are *x* test cases covered by the present suite, named `Test_01, ... , Test_x` (extending the base class `PIVX_FakeStakeTest`) which are run from the main script `pivx_fake_stake.py`

##### Test_00

The first test simply benchmarks the disk size allocation and verifies that the proposed fix does not filter out valid proof of stake blocks.
The node generates the first 250 proof of work blocks and then uses the collected rewards to stake the successive 15 blocks which are then relayed to the network.

##### Test_01

Covers a naive scenario representing the basic attack described in the article: "forked" blocks with a valid proof of stake but with an invalid (already spent on mainchain before the split) coinstake input, are spammed to the network.

Specifically, here, a node mines 250 blocks, then spends the mined coinbase outputs performing the “stake amplification” step of the attack. (other blocks are mined to confirm the spending transactions).
All the outputs are then used as coinstake inputs to create 15 spam blocks at a random heights in the range (N-20, N) where N is the current block height and those blocks are then sent from a node to his peers.

##### Test_02

Verifies that the proposed fix doesn’t block “valid” forked chains.
What might happen, in fact, is that some coins could be spent on main chain in a block which is higher than the “forked” one that contains  a coinstake transaction using the same coins as inputs.

The test node mines 150 blocks collecting the rewards and then mines 50 more blocks. Then spends the  coins mined in the first 150 blocks. Then mines 10 more blocks to include the transactions in the chain.
The same outputs mined in blocks 1-150 (and spent from block 201 onward) are used to create 15 spam blocks at a depth  between 151 and 200.

##### Test_03

Covers the following attack: "forked" blocks with a valid zPoS but with an invalid (already spent on mainchain before the split) coinstake input that actually is a zerocoin spend, serial double-spent.

In other words, performs the same attack described in Test_01 but using already spent zerocoins to create fake zPoS stakes.

##### Test_04

Performs the same check as in Test_02 verifying that zPoS forked blocks that stake a zerocoin which is spent on mainchain on an higher block are still accepted.

##### Test_05

Covers the scenario of a valid PoS block with a valid coinstake input that double spent the coinstake input in one of the transactions in the same block. 

-------------------------------

Sorry if I'm missing something, this has been a long work.

Awesome team work @Fuzzbawls & @random-zebra 🚀 .

-------------------------------

#### ---> And btw, this is still a WIP..

### CODE REVIEW

**Comment by Fuzzbawls** [comment]
ACK fced4e2

Note: not all the inherited test cases are fully functional, however the pressing need is to test against the new "fake stake" crap, which this does (and also provides for a much more updated regtest base to build on in the future).

Also Note: in a future PR, need to integrate the fake stake tests into the greater `test_runner.py` script.
